File: src/hyperhandler/client/exchange.py
"""Exchange API client for Hyperliquid."""

from decimal import Decimal
from typing import Any

from hyperhandler.client.base import BaseClient
from hyperhandler.client.order_builder import OrderBuilder
from hyperhandler.config import NetworkConfig
from hyperhandler.models import OrderResult, OrderStatus, OrderType, TradingSignal
from hyperhandler.signer import Signer


class ExchangeClient(BaseClient):
    """Client for Hyperliquid Exchange API (trading operations)."""

    def __init__(
        self,
        network: NetworkConfig,
        signer: Signer,
        slippage: Decimal = Decimal("0.005"),
        **kwargs,
    ):
        """Initialize the exchange client.

        Args:
            network: Network configuration.
            signer: EIP-712 signer.
            slippage: Default slippage for market orders.
        """
        super().__init__(network, **kwargs)
        self.signer = signer
        self.order_builder = OrderBuilder(slippage=slippage)

    @property
    def address(self) -> str:
        """Get the signer's address."""
        return self.signer.address

    async def place_order(
        self,
        asset_index: int,
        is_buy: bool,
        size: Decimal,
        price: Decimal,
        order_type: OrderType = OrderType.LIMIT,
        reduce_only: bool = False,
        vault_address: str | None = None,
    ) -> OrderResult:
        """Place a single order.

        Args:
            asset_index: Asset index from API.
            is_buy: True for buy, False for sell.
            size: Order size.
            price: Order price (with slippage for market).
            order_type: Order type.
            reduce_only: Whether this is a reduce-only order.
            vault_address: Optional vault address for vault trading.

        Returns:
            OrderResult with order status.
        """
        tif = "Ioc" if order_type == OrderType.MARKET else "Gtc"

        action = {
            "type": "order",
            "orders": [
                {
                    "a": asset_index,
                    "b": is_buy,
                    "p": OrderBuilder._format_price(price),
                    "s": OrderBuilder._format_size(size),
                    "r": reduce_only,
                    "t": {"limit": {"tif": tif}},
                }
            ],
            "grouping": "na",
        }

        return await self._execute_order(action, vault_address)

    async def place_order_from_signal(
        self,
        signal: TradingSignal,
        asset_index: int,
        current_price: Decimal | None = None,
        vault_address: str | None = None,
        sz_decimals: int = 0,
    ) -> list[OrderResult]:
        """Place orders from a trading signal.

        Args:
            signal: The trading signal.
            asset_index: Asset index from API.
            current_price: Current market price (required for market orders).
            vault_address: Optional vault address.
            sz_decimals: Size decimals for the asset (affects price rounding).

        Returns:
            List of OrderResult for each order (entry, SL, TP).
        """
        action = self.order_builder.build_order_payload(
            signal=signal,
            asset_index=asset_index,
            current_price=current_price,
            sz_decimals=sz_decimals,
        )

        result = await self._execute_order(action, vault_address)

        # For grouped orders, we get a single result
        # but we should return one result per order in the group
        num_orders = len(action.get("orders", []))
        if result.success and num_orders > 1:
            # Create placeholder results for SL/TP
            results = [result]
            for _ in range(num_orders - 1):
                results.append(
                    OrderResult(
                        success=True,
                        status=OrderStatus.OPEN,
                    )
                )
            return results

        return [result]

    async def _execute_order(
        self,
        action: dict[str, Any],
        vault_address: str | None = None,
    ) -> OrderResult:
        """Execute an order action.

        Args:
            action: The order action payload.
            vault_address: Optional vault address.

        Returns:
            OrderResult.
        """
        try:
            if vault_address:
                payload = self.signer.sign_action_for_vault(action, vault_address)
            else:
                payload = self.signer.sign_action(action)

            result = await self._post("exchange", payload)

            # Parse response
            if result.get("status") == "ok":
                response = result.get("response", {})
                data = response.get("data", {})

                # Extract order details if available
                statuses = data.get("statuses", [])
                if statuses:
                    first_status = statuses[0]
                    if isinstance(first_status, dict):
                        if "filled" in first_status:
                            filled = first_status["filled"]
                            return OrderResult(
                                success=True,
                                order_id=filled.get("oid"),
                                filled_size=Decimal(str(filled.get("totalSz", "0"))),
                                avg_price=Decimal(str(filled.get("avgPx", "0")))
                                if filled.get("avgPx")
                                else None,
                                status=OrderStatus.FILLED,
                            )
                        elif "resting" in first_status:
                            resting = first_status["resting"]
                            return OrderResult(
                                success=True,
                                order_id=resting.get("oid"),
                                status=OrderStatus.OPEN,
                            )
                        elif "error" in first_status:
                            return OrderResult(
                                success=False,
                                error=first_status["error"],
                                status=OrderStatus.REJECTED,
                            )

                return OrderResult(success=True, status=OrderStatus.PENDING)

            # Error response
            return OrderResult(
                success=False,
                error=result.get("response", "Unknown error"),
                status=OrderStatus.REJECTED,
            )

        except Exception as e:
            return OrderResult(
                success=False,
                error=str(e),
                status=OrderStatus.REJECTED,
            )

    async def cancel_order(
        self,
        asset_index: int,
        order_id: int,
        vault_address: str | None = None,
    ) -> bool:
        """Cancel an order.

        Args:
            asset_index: Asset index.
            order_id: Order ID to cancel.
            vault_address: Optional vault address.

        Returns:
            True if cancelled successfully.
        """
        action = self.order_builder.build_cancel_payload(asset_index, order_id)

        if vault_address:
            payload = self.signer.sign_action_for_vault(action, vault_address)
        else:
            payload = self.signer.sign_action(action)

        try:
            result = await self._post("exchange", payload)
            return result.get("status") == "ok"
        except Exception:
            return False

    async def cancel_all_orders(
        self,
        asset_index: int | None = None,
        vault_address: str | None = None,
    ) -> int:
        """Cancel all open orders.

        Args:
            asset_index: Optional asset index to filter. None for all.
            vault_address: Optional vault address.

        Returns:
            Number of orders cancelled.
        """
        # This would need to get open orders first, then cancel each
        # For now, we'll implement a basic version
        action = {
            "type": "cancelByCloid",
            "cancels": [],  # Empty means cancel all
        }

        if vault_address:
            payload = self.signer.sign_action_for_vault(action, vault_address)
        else:
            payload = self.signer.sign_action(action)

        try:
            result = await self._post("exchange", payload)
            if result.get("status") == "ok":
                return result.get("response", {}).get("data", {}).get("cancelled", 0)
        except Exception:
            pass
        return 0

    async def set_leverage(
        self,
        asset_index: int,
        leverage: int,
        is_cross: bool = True,
        vault_address: str | None = None,
    ) -> bool:
        """Set leverage for an asset.

        Args:
            asset_index: Asset index.
            leverage: Leverage value.
            is_cross: True for cross margin, False for isolated.
            vault_address: Optional vault address.

        Returns:
            True if successful.
        """
        action = self.order_builder.build_leverage_payload(
            asset_index=asset_index,
            leverage=leverage,
            is_cross=is_cross,
        )

        if vault_address:
            payload = self.signer.sign_action_for_vault(action, vault_address)
        else:
            payload = self.signer.sign_action(action)

        try:
            result = await self._post("exchange", payload)
            return result.get("status") == "ok"
        except Exception:
            return False

    async def close_position(
        self,
        asset_index: int,
        size: Decimal,
        is_long: bool,
        price: Decimal,
        vault_address: str | None = None,
    ) -> OrderResult:
        """Close a position.

        Args:
            asset_index: Asset index.
            size: Position size to close.
            is_long: True if closing a long position.
            price: Execution price (with slippage).
            vault_address: Optional vault address.

        Returns:
            OrderResult.
        """
        # To close, we place an opposite order with reduce_only
        return await self.place_order(
            asset_index=asset_index,
            is_buy=not is_long,  # Opposite direction
            size=size,
            price=price,
            order_type=OrderType.MARKET,
            reduce_only=True,
            vault_address=vault_address,
        )

File: internal/client/exchange.go
package client

import (
	"context"
	"encoding/json"
	"time"

	"github.com/shopspring/decimal"

	"github.com/skilus/hyperhandler/internal/config"
	"github.com/skilus/hyperhandler/internal/models"
	"github.com/skilus/hyperhandler/internal/order"
	"github.com/skilus/hyperhandler/internal/signer"
)

// ExchangeClient drives the Hyperliquid Exchange API (trading). Mirrors
// client/exchange.py:ExchangeClient.
type ExchangeClient struct {
	*BaseClient
	signer  *signer.Signer
	builder *order.Builder

	// nonce returns the action nonce (timestamp in ms). Overridable for tests.
	nonce func() int64
}

// NewExchangeClient builds an ExchangeClient with the given signer and slippage.
func NewExchangeClient(network config.NetworkConfig, s *signer.Signer, slippage decimal.Decimal, opts ...Option) *ExchangeClient {
	return &ExchangeClient{
		BaseClient: NewBaseClient(network, opts...),
		signer:     s,
		builder:    order.NewBuilder(slippage),
		nonce:      func() int64 { return time.Now().UnixMilli() },
	}
}

// Address returns the signer's address.
func (c *ExchangeClient) Address() string { return c.signer.Address() }

// signAction signs an action with a fresh nonce and the optional vault address.
func (c *ExchangeClient) signAction(action any, vaultAddress *string) (signer.Payload, error) {
	return c.signer.SignAction(action, c.nonce(), vaultAddress, nil)
}

// PlaceOrder places a single order. price is the wire price (caller applies
// slippage for market orders). Mirrors exchange.py:place_order.
func (c *ExchangeClient) PlaceOrder(
	ctx context.Context,
	assetIndex int,
	isBuy bool,
	size, price decimal.Decimal,
	orderType models.OrderType,
	reduceOnly bool,
	vaultAddress *string,
) models.OrderResult {
	action := c.builder.BuildSingleOrderPayload(assetIndex, isBuy, size, price, orderType, reduceOnly)
	return c.executeOrder(ctx, action, vaultAddress)
}

// PlaceOrderFromSignal builds and submits the order group for a signal (entry +
// optional SL/TP). On a successful grouped submit it returns one result per leg,
// with the SL/TP legs as OPEN placeholders. Mirrors exchange.py:place_order_from_signal.
func (c *ExchangeClient) PlaceOrderFromSignal(
	ctx context.Context,
	sig *models.TradingSignal,
	assetIndex int,
	currentPrice *decimal.Decimal,
	vaultAddress *string,
	szDecimals int,
) ([]models.OrderResult, error) {
	action, err := c.builder.BuildOrderPayload(sig, assetIndex, currentPrice, szDecimals)
	if err != nil {
		return nil, err
	}

	result := c.executeOrder(ctx, action, vaultAddress)

	numOrders := 1
	if orders, ok := action.Get("orders"); ok {
		if list, ok := orders.([]any); ok {
			numOrders = len(list)
		}
	}

	if result.Success && numOrders > 1 {
		results := []models.OrderResult{result}
		for i := 1; i < numOrders; i++ {
			results = append(results, models.OrderResult{Success: true, Status: models.StatusOpen})
		}
		return results, nil
	}
	return []models.OrderResult{result}, nil
}

// orderResponse is the /exchange success envelope for order/cancel actions.
type orderResponse struct {
	Status   string `json:"status"`
	Response struct {
		Data struct {
			Statuses []orderStatusEntry `json:"statuses"`
		} `json:"data"`
	} `json:"response"`
}

// orderStatusEntry is one entry of the statuses array; exactly one of the
// pointers is set per entry.
type orderStatusEntry struct {
	Filled  *filledStatus  `json:"filled"`
	Resting *restingStatus `json:"resting"`
	Error   *string        `json:"error"`
}

type filledStatus struct {
	Oid     int64            `json:"oid"`
	TotalSz decimal.Decimal  `json:"totalSz"`
	AvgPx   *decimal.Decimal `json:"avgPx"`
}

type restingStatus struct {
	Oid int64 `json:"oid"`
}

// executeOrder signs and posts an order action and parses the result. Any error
// (signing, transport, mapped HL error) is folded into a rejected OrderResult,
// matching the Python try/except that never propagates. Mirrors exchange.py:_execute_order.
func (c *ExchangeClient) executeOrder(ctx context.Context, action any, vaultAddress *string) models.OrderResult {
	payload, err := c.signAction(action, vaultAddress)
	if err != nil {
		return models.OrderResult{Success: false, Error: err.Error(), Status: models.StatusRejected}
	}

	raw, err := c.post(ctx, "exchange", payload, true)
	if err != nil {
		return models.OrderResult{Success: false, Error: err.Error(), Status: models.StatusRejected}
	}

	var resp orderResponse
	if err := json.Unmarshal(raw, &resp); err != nil {
		return models.OrderResult{Success: false, Error: err.Error(), Status: models.StatusRejected}
	}

	if resp.Status != "ok" {
		return models.OrderResult{Success: false, Error: "Unknown error", Status: models.StatusRejected}
	}

	statuses := resp.Response.Data.Statuses
	if len(statuses) > 0 {
		st := statuses[0]
		switch {
		case st.Filled != nil:
			oid := st.Filled.Oid
			return models.OrderResult{
				Success:    true,
				OrderID:    &oid,
				FilledSize: st.Filled.TotalSz,
				AvgPrice:   st.Filled.AvgPx,
				Status:     models.StatusFilled,
			}
		case st.Resting != nil:
			oid := st.Resting.Oid
			return models.OrderResult{Success: true, OrderID: &oid, Status: models.StatusOpen}
		case st.Error != nil:
			return models.OrderResult{Success: false, Error: *st.Error, Status: models.StatusRejected}
		}
	}

	return models.OrderResult{Success: true, Status: models.StatusPending}
}

// statusOK reports whether the raw /exchange response has status "ok".
func statusOK(raw []byte) bool {
	var env struct {
		Status string `json:"status"`
	}
	return json.Unmarshal(raw, &env) == nil && env.Status == "ok"
}

// CancelOrder cancels a single order. Returns false on any error, matching the
// Python except-swallow. Mirrors exchange.py:cancel_order.
func (c *ExchangeClient) CancelOrder(ctx context.Context, assetIndex int, orderID int64, vaultAddress *string) bool {
	action := c.builder.BuildCancelPayload(assetIndex, orderID)
	payload, err := c.signAction(action, vaultAddress)
	if err != nil {
		return false
	}
	raw, err := c.post(ctx, "exchange", payload, true)
	if err != nil {
		return false
	}
	return statusOK(raw)
}

// CancelAllOrders cancels all open orders (empty cancels = cancel all) and
// returns the number cancelled, 0 on any error. Mirrors exchange.py:cancel_all_orders.
func (c *ExchangeClient) CancelAllOrders(ctx context.Context, vaultAddress *string) int {
	action := signer.NewOrderedMap(
		"type", "cancelByCloid",
		"cancels", []any{},
	)
	payload, err := c.signAction(action, vaultAddress)
	if err != nil {
		return 0
	}
	raw, err := c.post(ctx, "exchange", payload, true)
	if err != nil {
		return 0
	}
	var resp struct {
		Status   string `json:"status"`
		Response struct {
			Data struct {
				Cancelled int `json:"cancelled"`
			} `json:"data"`
		} `json:"response"`
	}
	if json.Unmarshal(raw, &resp) != nil || resp.Status != "ok" {
		return 0
	}
	return resp.Response.Data.Cancelled
}

// SetLeverage sets the leverage for an asset. Mirrors exchange.py:set_leverage.
func (c *ExchangeClient) SetLeverage(ctx context.Context, assetIndex, leverage int, isCross bool, vaultAddress *string) bool {
	action := c.builder.BuildLeveragePayload(assetIndex, leverage, isCross)
	payload, err := c.signAction(action, vaultAddress)
	if err != nil {
		return false
	}
	raw, err := c.post(ctx, "exchange", payload, true)
	if err != nil {
		return false
	}
	return statusOK(raw)
}

// ClosePosition closes a position by placing an opposite reduce-only market
// order. Mirrors exchange.py:close_position.
func (c *ExchangeClient) ClosePosition(ctx context.Context, assetIndex int, size decimal.Decimal, isLong bool, price decimal.Decimal, vaultAddress *string) models.OrderResult {
	return c.PlaceOrder(ctx, assetIndex, !isLong, size, price, models.Market, true, vaultAddress)
}
